File: Data Structures/Doubly Linked List/doublylinkedlist.cpp
#include <bits/stdc++.h>
using namespace std;
class Node
{
public:
    int data;
    Node *prev;
    Node *next;
};
Node *head;
Node *getNewNode(int data)
{
    Node *newNode = new Node();
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = NULL;
    return newNode;
}
void printListWithAddresses()
{
    Node *temp = head;
    while (temp)
    {
        if (temp == head)
        {
            cout << "[Head " << temp->prev << "|" << temp << "|" << temp->data << "|" << temp->next << "]";
        }
        else if (temp->next == NULL)
        {
            cout << "-> [Tail: " << temp->prev << "|" << temp << "|" << temp->data << "|" << temp->next << "]";
        }
        else
        {
            cout << "-> [" << temp->prev << "|" << temp << "|" << temp->data << "|" << temp->next << "]";
        }
        temp = temp->next;
    }
    cout << endl;
}
void printListWithoutAddresses()
{
    Node *temp = head;
    while (temp)
    {
        if (temp == head)
        {
            cout << "[Head: " << temp->data << "]";
        }
        else if (temp->next == NULL)
        {
            cout << "-> [Tail: " << temp->data << "]";
        }
        else
        {
            cout << "-> [" << temp->data << "]";
        }
        temp = temp->next;
    }
    cout << endl;
}
void insertAt1(int data)
{
    Node *newNode = getNewNode(data);
    if (head == NULL)
    {
        head = newNode;
        printListWithoutAddresses();
        return;
    }
    Node *newNode1 = head;
    newNode1->prev = newNode;
    newNode->next = head;
    head = newNode;
    printListWithoutAddresses();
}
void insertAtLast(int data)
{
    Node *newNode2 = getNewNode(data);
    Node *newNode = head;
    while (newNode->next)
    {
        newNode = newNode->next;
    }
    newNode2->prev = newNode;
    newNode->next = newNode2;
    newNode2->next = NULL;
    printListWithoutAddresses();
}
void insertAtN(int data, int position)
{
    Node *newNode2 = getNewNode(data);
    Node *newNode = head;
    for (int i = 0; i < position - 2; i++)
    {
        newNode = newNode->next;
    }
    newNode2->next = newNode->next;
    newNode2->prev = newNode;
    newNode->next->prev = newNode2;
    newNode->next = newNode2;
    printListWithoutAddresses();
}
void print_reverse_list()
{
    Node *newNode = head;
    while (newNode->next)
    {
        newNode = newNode->next;
    }
    while (newNode)
    {
        cout << newNode->data << " ";
        newNode = newNode->prev;
    }
}

int main()
{
    head = NULL;
    insertAt1(20);    // 20
    insertAtLast(30); // 20 30
    insertAt1(40);    // 40 20 30
    insertAtLast(50); // 40 20 30 50
    insertAtLast(60); // 40 20 30 50 60
    insertAtN(70, 3); // 40 20 70 30 50 60
    insertAtN(80, 5); // 40 20 70 30 80 50 60
    print_reverse_list();
}

File: Data Structures/Doubly Linked List/doublylinkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None

    def __len__(self):
        new_node = self.head
        count = 1
        while new_node.next:
            count += 1
            new_node = new_node.next
        return count

    def __repr__(self):
        new_node = self.head
        nodes = []
        while new_node:
            if new_node is self.head:
                nodes.append(f"[Head: {new_node.data}]")
            elif new_node.next is None:
                nodes.append(f"[Tail: {new_node.data}]")
            else:
                nodes.append(f"[{new_node.data}]")
            new_node = new_node.next
        return f"{'-> '.join(nodes)} Length: {len(self)}"

    def insertAt1(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            print(self)
            return
        new_node1 = self.head
        new_node1.prev = new_node
        new_node.next = self.head
        self.head = new_node
        print(self)

    def insertAtLast(self, data):
        new_node = Node(data)
        new_node1 = self.head
        while new_node1.next:
            new_node1 = new_node1.next
        new_node1.next = new_node
        new_node.prev = new_node1
        print(self)

    def insertAtN(self, data, position):
        new_node = Node(data)
        new_node1 = self.head
        for i in range(position - 2):
            new_node1 = new_node1.next
        new_node.next = new_node1.next
        new_node.prev = new_node1
        new_node1.next.prev = new_node
        new_node1.next = new_node
        print(self)

    def print_reverse_list(self):
        new_node = self.head
        while new_node.next:
            new_node = new_node.next
        while new_node:
            print(new_node.data, end=" ")
            new_node = new_node.prev


if __name__ == "__main__":
    obj = DoublyLinkedList()
    obj.insertAt1(10)
    obj.insertAtLast(30)
    obj.insertAt1(201)
    obj.insertAtN(312, 3)
    obj.insertAtLast(34)
    obj.insertAtN(23, 2)
    obj.insertAtN(50, 5)
    obj.insertAt1(300)
    obj.print_reverse_list()
